### ISSUE TITLE
[v0.7.0] Multiple recipes with variants on the same JSX element bug - only the 1st recipe (as specified in the config) has its chosen variant's styles generated

### ISSUE THREAD

**Issue by thetiby** [chakra-ui/panda#1093]
### Description

Recipe extraction when multiple recipes are used on the same component doesn't work when selecting variants for those recipes.
 
I also included an example to better showcase the issue, alongside with a similar use case that suffers from the same bug.

### Link to Reproduction

https://stackblitz.com/edit/github-cggwkl-43k2mm?file=themePreset.ts,src%2Fapp%2Fpage.tsx,panda.css,panda.config.ts

### Steps to reproduce

1. Go to `themePreset.ts` and notice **the order** of 2 recipes that apply **to the same** 2 JSX elements (for the 2 use cases)
2. Go to `src/app/page.tsx` for the usage
3. Go to `panda.css` and notice the relevant styles (run `panda cssgen --outfile panda.css` to get a fresh file)

```css
@layer recipes {
  .recipe1--borderStyle_orange {
    border: 10px solid orange
    }

  .recipe1--borderStyle_coral {
    border: 10px solid coral
    }

  .recipe2--bgStyle_lightgreen {
    background: lightgreen
    }

  @layer _base {
    .recipe1 {
      padding: 30px
        }

    .recipe1,.recipe2 {
      display: flex
        }
    }
}
``` 
4. The next app suffers from the same issue (using the PostCSS plugin), as it can be noticed:
![image](https://github.com/chakra-ui/panda/assets/2418978/8b5b6d85-0182-454d-af34-8658c0e52a63)

### JS Framework

React (TS)

### Panda CSS Version

0.7.0

### Browser

Google Chrome Version 114.0

### Operating System

- [X] macOS
- [ ] Windows
- [ ] Linux

### Additional Information

[There is a fix](https://github.com/chakra-ui/panda/pull/879) that addressed an older issue, where multiple recipes did not work at all on a single JSX element. However, as we can see [in the tests for that fix](https://github.com/chakra-ui/panda/pull/879/files#diff-574b47b98600b2117540eaa3e43de60d9607b477fc93a8fbdcf0b5d73edde332R289), only the base CSS styles (the `color` property) were addressed, missing `fontSize` property of the `variant=small` 

Thank you for taking the time and for your quick turnaround!

**Comment by astahmer** [comment]
Hey, the test case was actually wrong ! thank you for pointing it out, I fixed it with Sage [in that PR](https://github.com/chakra-ui/panda/pull/1100/files#diff-574b47b98600b2117540eaa3e43de60d9607b477fc93a8fbdcf0b5d73edde332R206)

then, in the same PR we also added the missing `{recipe}.raw` (that I forgot to add), which solves your use-case.

instead of:
```tsx
<TestReactComponentParent borderProps={{
        borderStyle: 'orange'
      }} bgProps={{
        bgStyle: 'yellow'
      }}>
```

you can do:

```tsx
<TestReactComponentParent
        borderProps={recipe1.raw({ borderStyle: "orange" })}
        bgProps={recipe2.raw({ bgStyle: "yellow" })}
```

**Comment by thetiby** [comment]
Hello @astahmer, thank you for the speedy updates!
I am struggling to find docs for `raw`, so I am not sure the above example does what I think it does.

If you don't mind, would you clarify some questions about it?

1. In the tests you fixed, `raw` isn't used, just `<ComponentWithMultipleRecipes variant="small" />`, and yet, all of the variations are generated, as opposed to the example linked above. Is it because all of the recipes use the same name for the variant? what if the 2nd recipe had a different variant, called `colorVariant`, is `<ComponentWithMultipleRecipes variant="small" colorVariant="blue" />`? 

In the example linked above, even the basic case `<div className={cx(recipe1({ borderStyle }), recipe2({ bgStyle }))}>{children}</div>` failed, when using `<TestReactComponent borderStyle='orange' bgStyle='yellow'>`.

2. adding the JSX component name to the recipe seemed to be the point in order to avoid having to use the recipe for each usage of the comonent and only pass the props instead; the `raw` example above seems to do the same as
```ts
      <TestReactComponentParent borderProps={{
        borderStyle: recipe1({ borderStyle: 'orange' })
      }} bgProps={{
        bgStyle: recipe2({ bgStyle: 'yellow'})
      }}>
        The custom variants are generated, as the recipe is used directly, not via props.
      </TestReactComponentParent >
```
but less verbose; it even works without the `jsx` hint ([`TestReactComponentParentWithoutRecipe` is not specified in the recipes](https://stackblitz.com/edit/github-cggwkl-43k2mm?file=themePreset.ts)) - the screenshot bellow also coincidentally shows the 1st and 2nd examples as they are expected to work, since the 3rd hardcoded one generates the missing styles for them
<img width="1393" alt="image" src="https://github.com/chakra-ui/panda/assets/2418978/bbd844f3-8478-4585-ab47-56ed7aa67b3e">

**Comment by segunadebayo** [comment]
Closing this out since we've arrived at a good solution.

- Either apply the recipe at the prop level, or
- Use the upcoming `<recipe>.raw` syntax (will be documented pretty soon)

**Comment by thetiby** [comment]
@segunadebayo  
Could you please provide a quick example with the 2 good solutions you mentioned? (by simply forking the one I provided)

It's hard to understand how to apply these 2 good solutions.

Thank you!

**Comment by astahmer** [comment]
>  Is it because all the recipes use the same name for the variant? What if the 2nd recipe had a different variant, called colorVariant, is <ComponentWithMultipleRecipes variant="small" colorVariant="blue" />?

yes & yes, if you keep the same property name while tagging the component as part of the jsx for that recipe, we interpret that prop/value as part of that recipe in JSX too !

---

> Could you please provide a quick example with the 2 good solutions you mentioned? (by simply forking the one I provided)
yes ->

> Either apply the recipe at the prop level, or

this one is what you had here:
```tsx
<TestReactComponent borderStyle='orange' bgStyle='yellow'>
```

I did notice an issue remaining with this form, which led to [this fix](https://github.com/chakra-ui/panda/pull/1097/commits/a0d42c831f7383c7cf9e54dc1b1c5f3e1cfd237d) (commit including a failing test so we can see the diff after the fix [here](https://github.com/chakra-ui/panda/pull/1097/commits/1321edc95ffd7b7dc25d38632b8691f08b28141a))

---

> Use the upcoming <recipe>.raw syntax (will be documented pretty soon)

this one is the one I wrote above:

```tsx
<TestReactComponentParent
        borderProps={recipe1.raw({ borderStyle: "orange" })}
        bgProps={recipe2.raw({ bgStyle: "yellow" })}
```

**Comment by thetiby** [comment]
Thank you very much, @astahmer!

Yes, releasing [`fix: recipes.splitProps`](https://github.com/chakra-ui/panda/pull/1097/commits/95398507a6dabd9af10e36b12b3318c27b108cf8)  and  the one you mentioned above ([`fix(recipes): splitProps filter style props using other recipes variant keys`](https://github.com/chakra-ui/panda/pull/1097/commits/a0d42c831f7383c7cf9e54dc1b1c5f3e1cfd237d)) will definitely fix this usecase:
```
<TestReactComponent borderStyle='orange' bgStyle='yellow'>
```
---

If the question below will be captured in the docs, please just link to those docs once they are available, no need to answer here as well.

What is the advantage of using: 
```jsx
<TestReactComponentParent
        borderProps={recipe1.raw({ borderStyle: "orange" })}
        bgProps={recipe2.raw({ bgStyle: "yellow" })}
```
over the currently working (albeit more verbose)
```jsx
<TestReactComponentParent
        borderProps={{
          borderStyle: recipe1({ borderStyle: 'orange' })
        }} 
        bgProps={{
          bgStyle: recipe2({ bgStyle: 'yellow'})
        }}
```

Thank you again for taking the time!

**Comment by thetiby** [comment]
@segunadebayo  
Hey there, would you reopen this one please, to track the release of the upcoming `.raw` syntax and [the above mentione fixes](https://github.com/chakra-ui/panda/pull/1097/commits/95398507a6dabd9af10e36b12b3318c27b108cf8), since, currently, none [of these 2 solutions](https://github.com/chakra-ui/panda/issues/1093#issuecomment-1649380094) are available as described in this issue.

Thank you!

**Comment by astahmer** [comment]
re-opening cause of [this commit](https://github.com/chakra-ui/panda/pull/1097/commits/5d4d8768ee884f98b5551d6f42d1685d3fa4e949) as you said but I think the `recipe.raw` has been merged & published yesterday so this might work fine (havent re-tested):

```tsx
<TestReactComponentParent
        borderProps={recipe1.raw({ borderStyle: "orange" })}
        bgProps={recipe2.raw({ bgStyle: "yellow" })}
```

**Comment by segunadebayo** [comment]
The `recipe.raw` syntax works now. I also recommend using the `staticCss` option to pre-generate styles for dynamic scenarios like this.

These complex use cases can be greatly simplified using the basic primitives (css, cva). I'm glad we now provide tools to manage these edge cases. We don't want to over-engineer the compiler beyond what we currently have.

Thanks for bringing this up.

**Comment by thetiby** [comment]
@segunadebayo @astahmer 

I can confirm that the problem is fixed ✅ with the `v0.9.0` release, thank you for your quick and helpful responses! ✨⚡️

<img width="1742" alt="image" src="https://github.com/chakra-ui/panda/assets/2418978/7f6131ce-8f5c-4459-9e74-f8775cd8e97a">

---

On `recipe.raw`, I am not sure if this is intended or if it is not, but I noticed that it works similar to using `staticCss`; if `recipe.raw` is used _anywhere in the code_, the corresponding recipe styles will be generated, _regardless of the control flow_ (for example; `if (false) recipe2.raw({ bgStyle: "yellow" })` will generate the bgStyle--yellow variant even if it's not actually used)

Obviously, this example has no real-world usage, but a more advanced one, where the flow can be statically determined, could make use of `recipe.raw` if it did follow that flow.

I'm going to keep an eye for more examples / docs of usage for `recipe.raw`, as I think I may have misunderstood its purpose.

Thank you again for your hard work! 💫

**Comment by astahmer** [comment]
> On recipe.raw, I am not sure if this is intended or if it is not, but I noticed that it works similar to using staticCss; if recipe.raw is used anywhere in the code, the corresponding recipe styles will be generated, regardless of the control flow

yes it is intended and it acts more like `css()` or any recipe/pattern function, they also do not depend on control flow

having this would always generate both blue.100 and red.100

```tsx
const App = () => {
    return true ? css({ color: "blue.100" }) : css({ color: "red.100" })
}
```
